middleware.js

```diff
@@ -0,0 +1,211 @@
+ * | '_  \| | | || | | |
+ * | | | || |_| || | | |
+ * |_| |_|\___,_||_| |_|
+ *
+ * (c) Huli Inc
+ */
+
+
+/**
+ * @file Manages and executes middlewares in all the different app phases
+ * @requires ju-shared/observable-class
+ * @module ju-mvc/middleware
+ * @extends ju-shared/observable-class
+ */
+
+define([
+	'ju-shared/observable-class'
+],
+function (
+	ObservableClass
+){
+	'use strict';
+
+
+	/** Constants **/
+
+    /**
+     * @constant {String} DEFAULT_SUBPHASE - if the subphase is not provided when adding a middleware this will be use
+     */
+	var DEFAULT_SUBPHASE = 'during';
+
+    /**
+     * Object used to store all the phases and middlewares
+     * @type {Object}
+     */
+    var phases = {};
+
+
+    /**
+     * Create an instance of middleware
+     * @constructor
+     * @alias module:ju-mvc/middleware
+     * @returns {Object} Middleware
+     */
+	var Middleware = ObservableClass.extend({
+
+
+		init : function(){
+            //initialize the phases
+            var pashesKeys = Object.keys(Middleware.PHASES);
+            for (var currentPhase = 0, pashesTotal = pashesKeys.length; currentPhase < pashesTotal; currentPhase++) {
+                var phaseName = Middleware.PHASES[pashesKeys[currentPhase]];
+                if(!phases[phaseName]){
+                    phases[phaseName] = {
+                        before : [],
+                        during : [],
+                        after : []
+                    };
+                }
+            }
+		},
+
+        /**
+         * Add a middleware to a specified phase
+         * @param {Middleware} middleware
+         * @param {String} phase - check Middleware.PHASES
+         * @param {String} subPhase - check Middleware.SUBPHASES
+         * @returns {boolean}
+         */
+		add : function (middleware, phase, subPhase){
+			if (phases[phase]){
+				if (phases[phase][subPhase]){
+					phases[phase][subPhase].push(middleware);
+				}
+				else{
+					phases[phase][DEFAULT_SUBPHASE].push(middleware);
+				}
+                return true;
+			}else {
+                return false;
+            }
+		},
+
+        /**
+         * Executes the middlewares sequentially for the given phase/subphase
+         * @param {String} phase - phase to be executed, check Middleware.PHASES
+         * @param {String} subPhase - subphase to be executed, check Middleware.SUBPHASES
+         * @param {Object} params - params to be pass to all the middlewares
+         * @param {Function} [onSuccess] - callback success function
+         * @param {Function} [onError] - callback error function
+         */
+		run : function(phase, subPhase, params, onSuccess, onError) {
+			if (phases[phase] && phases[phase][subPhase]){
+				this._runAll(phases[phase][subPhase], params).then(function (result) {
+					log('Middleware: ran successfully '+ phase + ':' + subPhase, result);
+					if (typeof onSuccess === 'function'){
+						onSuccess(result);
+					}
+				}).catch(function (error) {
```

Comment:
we usually use `['catch']` as some browsers have issues with `.catch` .. it might not be an issue anymore, but I think it's a good idea to use the same format for consistency